src/profile-logic/profile-data.js

```diff
@@ -683,20 +684,20 @@ export function invertCallstack(thread: Thread): Thread {
 
 export function getSampleIndexClosestToTime(
   samples: SamplesTable,
-  time: number
+  time: number,
+  interval: number
 ): IndexIntoSamplesTable {
-  // TODO: This should really use bisect. samples.time is sorted.
-  for (let i: number = 0; i < samples.length; i++) {
-    if (samples.time[i] >= time) {
-      if (i === 0) {
-        return 0;
-      }
-      const distanceToThis = samples.time[i] - time;
-      const distanceToLast = time - samples.time[i - 1];
-      return distanceToThis < distanceToLast ? i : i - 1;
-    }
+  // Bisect to find the index of the first sample after the provided time.
+  const index = bisection.right(samples.time, time);
```

Comment:
I think you need to adjust by half the interval here as well, otherwise you'll get wrong results if there are "overlapping" samples. For example:

`getSampleIndexClosestToTime([1, 1.3, 1.5, 2], 1.7, 1)` should give `1` (`1.3`), because 1.7 is closest to 1.8 which is the midpoint of the sample that starts at 1.3. Your function however will compare 1.7 to the samples at 1.5 and 2.

You can add a comment along the lines of
```
// We want to find the first sample whose center point is to the right of the clicked position.
// The center point of a sample is located at sample.time + interval / 2.
// Instead of adding interval / 2 to every sample time, we *subtract* it from `time`.
```

Reply:
I initially did that but I reverted it. I'll push an updated patch with your suggestion but I'll share my reasoning just in case.

Given the way the draw code works here, the 1.5 sample is drawn over the 1.3 sample. From a UI point of view, clicking at 1.7 should still select the 1.5 sample. That's why my intent was to always pick the last sample before a time, and compare it with the next one in case there was a gap.

To match a click, maybe we should actually only try to find the "last before" sample and return it if `sampleTime + interval > clickTime`.

Reply:
Oh I see, that makes sense! Ok, let's keep what you had, but with the additional `index == samples.length` check.

I don't think it would be a good idea to ignore clicks in empty areas. That can be very frustrating if the profile looks like the one I linked earlier, where there is a gap after every sample.